### PR TITLE
1891 Gossipsub RPC validation inspector false positive E2E testing

### PR DESCRIPTION
**Pull request by kc1116**
This PR adds an E2E test that covers false positive validation errors in the gossipsub RPC control message validation inspector. This is crucial since the current implementation relies on certain assumptions about GossipSub. These assumptions are based on our analysis of the GossipSub code, but the GossipSub team does not guarantee that it will hold in the long term. This PR leverages the fact that an invalid control message notification is distributed by the rpc inspector notification distributor, under normal operations no invalid control message notifications are expected to be disseminated.

### CODE REVIEW

File: integration/tests/bft/gossipsub/rpc_inspector/suite.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	for containerName, metric := range mets {
+		val := metric[0].GetGauge().GetValue()
+		require.Zerof(s.T(), val, fmt.Sprintf("expected inspector notification queue cache size for container %s to be 0 got %v", containerName, val))
+	}
+}
+
+// inspectorNotifQSizeMetricName returns the metric name for the rpc inspector notification queue cache size.
+func (s *Suite) inspectorNotifQSizeMetricName() string {
+	return fmt.Sprintf("network_hero_cache_%s_successful_write_count_total", metrics.ResourceNetworkingRpcInspectorNotificationQueue)
+}
```
Comment (yhassanzadeh13) [inline integration/tests/bft/gossipsub/rpc_inspector/suite.go:73]: The current measure inaccurately portrays the event or the detection of misbehavior. This is due to the metric displaying the count of _queued_ misbehavior notifications awaiting processing by the scoring component. If the queue worker operates rapidly enough, and there is no congestion in the queue, the notifications are processed instantaneously, leaving the queue perpetually empty. However, this absence of a queued notification doesn't necessarily mean there was no notification generation.

To resolve this discrepancy and correctly ascertain the event of "no notification detected", I propose two potential solutions:

1. (Preferred) Incorporate a counter metric within the `Distribute` [method](https://github.com/onflow/flow-go/blob/13383f1c08bad77e639930c846535dff84010653/network/p2p/distributor/gossipsub_inspector.go#L83). This metric should tally the aggregate number of notifications received by the inspectors. Subsequently, this test could verify the metric's value.

2. An alternate, albeit less conventional, solution also exists. At present, the [queue worker](https://github.com/onflow/flow-go/blob/13383f1c08bad77e639930c846535dff84010653/network/p2p/distributor/gossipsub_inspector.go#L20) is set to `1`. However, this parameter could be made available as a flag, which could be overridden to `0` during this test. This ensures that if a notification is ever queued, it remains unprocessed. Therefore, the current test format guarantees the capture of the event of "no notification detected".

**Review by yhassanzadeh13 — COMMENTED** [review]
I like the general approach that the test is taking, though a few concerns need to be addressed. Appreciate your consideration; thanks!

File: integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+type GossipsubRPCInstpectorFalePostiveNotificationsTestSuite struct {
+	Suite
+}
+
+func TestGossipsubRPCInstpectorFalePostiveNotifications(t *testing.T) {
+	suite.Run(t, new(GossipsubRPCInstpectorFalePostiveNotificationsTestSuite))
+}
+
+// TestGossipsubRPCInstpectorFalePostiveNotifications this test ensures that any underlying changes or updates to any of the underlying libp2p libraries
```
Comment (gomisha) [inline integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go:19]: multiple typos in function name at start of comment (`TestGossipsubRPCInstpectorFalePostiveNotifications` should be `TestGossipsubRPCInspectorFalsePositiveNotifications`) along with actual function name that was already flagged in https://github.com/onflow/flow-go/pull/4371/files#r1210747930

File: integration/tests/bft/gossipsub/rpc_inspector/suite.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+// indicating no notifications have been disseminated.
+func (s *Suite) ensureNoNotificationsDisseminated(mets map[string][]*io_prometheus_client.Metric) {
+	for containerName, metric := range mets {
+		val := metric[0].GetGauge().GetValue()
+		require.Zerof(s.T(), val, fmt.Sprintf("expected inspector notification queue cache size for container %s to be 0 got %v", containerName, val))
+	}
+}
+
+// inspectorNotifQSizeMetricName returns the metric name for the rpc inspector notification queue cache size.
+func (s *Suite) inspectorNotifQSizeMetricName() string {
```
Comment (gomisha) [inline integration/tests/bft/gossipsub/rpc_inspector/suite.go:71]:
```suggestion
func (s *Suite) inspectorNotificationQSizeMetricName() string {
```

File: integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go (hunk 2 above)
Comment (kc1116) [reply integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go]: https://github.com/onflow/flow-go/pull/4371/commits/05f793b0f611821b359d52c8d59644c01ddbccbb

File: integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// does not result in any of the gossip sub rpc control message inspector validation rules being broken. Anytime a validation rule is broken an invalid
+// control message notification is disseminated. Using this fact, this tests sets up a full flow network and submits some transactions to generate network
+// activity. After some time we ensure that no invalid control message notifications are disseminated.
+func (s *GossipsubRPCInstpectorFalePostiveNotificationsTestSuite) TestGossipsubRPCInstpectorFalePostiveNotifications() {
+	// the network has started submit some transactions to create flow accounts.
+	// We wait for each of these transactions to be sealed ensuring we generate
+	// some artificial network activity
+	for i := 0; i < numOfTestAccounts; i++ {
+		s.submitSmokeTestTransaction(s.Ctx)
+	}
```
[the remark this answers is not in the document]
Comment (kc1116) [reply integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go]: https://github.com/onflow/flow-go/pull/4371/commits/36787884c85ffa729f65af62362cd7ecb502c2f8

File: integration/tests/bft/base_suite.go [hunk 5]
```diff
@@ -102,7 +102,9 @@ func (b *BaseSuite) SetupSuite() {
 func (b *BaseSuite) TearDownSuite() {
 	b.Net.Remove()
 	b.Cancel()
-	unittest.RequireCloseBefore(b.T(), b.OrchestratorNetwork.Done(), 1*time.Second, "could not stop orchestrator network on time")
+	if b.OrchestratorNetwork != nil {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply integration/tests/bft/base_suite.go:106]: https://github.com/onflow/flow-go/pull/4371/commits/433e2666b45774a3c0eebb472524086acf444569

File: integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package rpc_inspector
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+)
+
+const numOfTestAccounts = 5
```
[the remark this answers is not in the document]
Comment (kc1116) [reply integration/tests/bft/gossipsub/rpc_inspector/false_positive_test.go]: https://github.com/onflow/flow-go/pull/4371/commits/c000dae212b2e03efc169e13431a9cfc960c331a

File: integration/tests/bft/gossipsub/rpc_inspector/suite.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+// waitStateCommitments waits for n number of state commitment changes.
+func (s *Suite) waitForStateCommitments(ctx context.Context, n int, waitFor, tick time.Duration) {
+	startStateComm := s.getCurrERFinalStateCommitment(ctx)
+	numOfStateCommChanges := 0
+	require.Eventually(s.T(), func() bool {
+		currStateComm := s.getCurrERFinalStateCommitment(ctx)
+		if startStateComm != currStateComm {
+			numOfStateCommChanges++
+		}
+		return numOfStateCommChanges == n
```
Comment (yhassanzadeh13) [inline integration/tests/bft/gossipsub/rpc_inspector/suite.go:112]:
```suggestion
		return numOfStateCommChanges >= n
```

File: integration/tests/bft/gossipsub/rpc_inspector/suite.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	}, waitFor, tick)
+}
+
+// getCurrentFinalizedHeight returns the current finalized height.
+func (s *Suite) getCurrERFinalStateCommitment(ctx context.Context) string {
```
Comment (yhassanzadeh13) [inline integration/tests/bft/gossipsub/rpc_inspector/suite.go:117]:
```suggestion
func (s *Suite) getCurrentFinalExecutionStateCommitment(ctx context.Context) string {
```

**Review by yhassanzadeh13 — APPROVED** [review]
🚀 thanks for developing the test. Except for one correction that I pointed out, the rest looks great!

**Review by gomisha — APPROVED** [review]
Thanks for addressing all my feedback.

File: integration/tests/bft/gossipsub/rpc_inspector/suite.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// waitStateCommitments waits for n number of state commitment changes.
+func (s *Suite) waitForStateCommitments(ctx context.Context, n int, waitFor, tick time.Duration) {
+	startStateComm := s.getCurrERFinalStateCommitment(ctx)
+	numOfStateCommChanges := 0
+	require.Eventually(s.T(), func() bool {
+		currStateComm := s.getCurrERFinalStateCommitment(ctx)
+		if startStateComm != currStateComm {
+			numOfStateCommChanges++
```
[the remark this answers is not in the document]
Comment (kc1116) [reply integration/tests/bft/gossipsub/rpc_inspector/suite.go:108]: https://github.com/onflow/flow-go/pull/4371/commits/d24265f5c39ba2bac2a8ceee2f0d31c5be18b943

**Pull request merged**